### PR TITLE
more ergonomic stream redirection

### PR DESCRIPTION
**Pull request by jw3126**
Usually when I want to redirect `stdout`, I also want to redirect `stderr`. And often I want to redirect them to some kind of logfile.
This is quite a bit of boilerplate:
```julia
open("stdout.log", "w") do new_stdout
     open("stderr.log", "w" do new_stderr
         redirect_stdout(new_stdout) do
             redirect_stderr(new_stderr) do
                   f()
             end
        end
    end
end
```
With this PR it is just
```julia
redirect_stdio(f, stdout="stdout.txt", stderr="stderr.txt")
```

### CODE REVIEW

**Comment by jw3126** [comment]
CI fail looks related to PR. However I cannot reproduce this error locally. Ideas?

**Comment by JeffBezanson** [comment]
What happens when you open the same file twice, and redirect stdout and stderr to the two descriptors?

**Comment by jw3126** [comment]
Ah, good spot. I need to make sure if `stdout=stderr=/same/path` that only a single `io` is used.

File: base/stream.jl [hunk 1]
```diff
@@ -1215,20 +1215,95 @@ i.e. data to be read from [`stdin`](@ref) may be written to `wr`.
...
+    same_path(x, y) = false
+    same_path(x::AbstractString, y::AbstractString) = x == y
+    if same_path(stderr, stdin)
+        throw(ArgumentError("stdin and stderr cannot be the same path"))
+    end
+    if same_path(stdout, stdin)
+        throw(ArgumentError("stdin and stdout cannot be the same path"))
+    end
+
+    if same_path(stderr, stdout)
```
Comment (jw3126) [inline base/stream.jl:1279]: @JeffBezanson If the user passes the same path for `stdout` and `stderr` the `io` is opened only once and everything works fine. 
If the user passes e.g. two distinct `io` that point to the same file, that is still a problem. But the same is true on master e.g:
```julia
single_path = "single_file.txt"
open(single_path, "w") do new_stdout
     open(single_path, "w" do new_stderr
         redirect_stdout(new_stdout) do
             redirect_stderr(new_stderr) do
                   f()
             end
        end
    end
end
```
Should I document this? Should I try to handle it (I think an `IOStream` remembers its path)?

Comment (jw3126) [reply base/stream.jl:1301]: I documented this.

File: base/stream.jl [hunk 2]
```diff
@@ -1215,20 +1215,118 @@ i.e. data to be read from [`stdin`](@ref) may be written to `wr`.
...
+function redirect(;stdin=nothing, stderr=nothing, stdout=nothing)
+    stdin  === nothing || redirect_stdin(stdin)
+    stderr === nothing || redirect_stderr(stderr)
+    stdout === nothing || redirect_stdout(stdout)
+end
+
+"""
+    redirect(f; stdin=nothing, stderr=nothing, stdout=nothing)
+
+
```
Comment (JeffBezanson) [inline base/stream.jl:1233]: extra line break

File: base/stream.jl [hunk 3]
```diff
@@ -1215,20 +1215,118 @@ i.e. data to be read from [`stdin`](@ref) may be written to `wr`.
...
+julia> read("stdout.txt", String)
+"hello stdout"
+
+julia> read("stderr.txt", String)
+"hello stderr"
+```
+
+# Edge cases
+
+It is possible to pass the same exactly same argument to `stdout` and `stderr`:
```
Comment (JeffBezanson) [inline base/stream.jl:1258]:
```suggestion
It is possible to pass the same argument to `stdout` and `stderr`:
```

File: base/stream.jl [hunk 4]
```diff
@@ -1215,20 +1215,118 @@ i.e. data to be read from [`stdin`](@ref) may be written to `wr`.
...
+
+"""
+    redirect(;stdin=stdin, stderr=stderr, stdout=stdout)
+
+Redirect a subset of the streams `stdin`, `stderr`, `stdout`.
+"""
+function redirect(;stdin=nothing, stderr=nothing, stdout=nothing)
+    stdin  === nothing || redirect_stdin(stdin)
+    stderr === nothing || redirect_stderr(stderr)
+    stdout === nothing || redirect_stdout(stdout)
```
Comment (JeffBezanson) [inline base/stream.jl:1227]: It seems to me the path and error checking should be moved here, so that you get them with or without a function argument?

Comment (jw3126) [reply base/stream.jl:1227]: Note that this function does not accept paths. The reason is, that while this function could open the corresponding file, it cannot close it. So there is no error hadling that can be done here.

File: base/stream.jl (hunk 2 above)
Comment (jw3126) [reply base/stream.jl]: done

**Comment by jw3126** [comment]
Can this be merged?

**Comment by StefanKarpinski** [comment]
Bump.

**Comment by jw3126** [comment]
@StefanKarpinski I am a bit surprised, that this needs triage. Anyway, I am glad that this PR gets some attention. How can I find out when the next triage session is?

**Comment by oscardssmith** [comment]
In general, any user facing change needs triage. This doesn't mean it's expected to be controversial, just that we don't want to be adding features carelessly. Triage meetings are (almost) every week at 14:30 EST on Thursdays. You're welcome to come, but don't feel like you need to.

**Comment by jw3126** [comment]
Bump! Is there interest in this? If so I can resolve the merge conflicts.

**Comment by StefanKarpinski** [comment]
Would be good to look at this on a triage call.

**Comment by vtjnash** [comment]
Needs rebase, but triage approves

**Comment by KristofferC** [comment]
`redirect` seems like a way too generic name for this. `redirect_standard_streams` or something would be better imo.

**Comment by jw3126** [comment]
> `redirect` seems like a way too generic name for this. `redirect_standard_streams` or something would be better imo.

I agree the name is too generic. `redirect_standard_streams` is a bit verbose though. I like `redirect_stdio`. There is a struct with this name
https://github.com/JuliaLang/julia/blob/ad97dbd1b034e478775a4334ce22db9d4ab5eafb/base/stream.jl#L1130-L1133

but I think that struct is a private implementation detail and lower case is odd for a type anyway. So the struct can be renamed. Sounds good?

File: base/stream.jl [hunk 5]
```diff
@@ -1227,9 +1231,118 @@ Note that the direction of the stream is reversed.
...
+```
+Also the `stdin` argument may not be the same descriptor as `stdout` or `stderr`.
+```julia
+julia> io = open(...)
+
+julia> redirect(f, stdout=io, stdin=io) # not supported
+```
+
+!!! compat "Julia 1.6"
+    `redirect` requires Julia 1.6 or later.
```
Comment (fredrikekre) [inline base/stream.jl:1301]:
```suggestion
!!! compat "Julia 1.7"
    `redirect` requires Julia 1.7 or later.
```

File: base/stream.jl [hunk 6]
```diff
@@ -1227,9 +1231,118 @@ Note that the direction of the stream is reversed.
...
+
+See also [`redirect`](@ref).
 """
 redirect_stdin
 
+"""
+    redirect(;stdin=stdin, stderr=stderr, stdout=stdout)
+
+Redirect a subset of the streams `stdin`, `stderr`, `stdout`.
+Each argument must be an `IOStream`, `TTY`, `Pipe`, socket, or `devnull`.
```
Comment (fredrikekre) [inline base/stream.jl:1243]: Add `!!! compat` to this too maybe?

File: NEWS.md [hunk 7]
```diff
@@ -46,6 +46,7 @@ New library functions
 * New functor `Returns(value)`, which returns `value` for any arguments ([#39794])
 * New macro `Base.@invoke f(arg1::T1, arg2::T2; kwargs...)` provides an easier syntax to call `invoke(f, Tuple{T1,T2}, arg1, arg2; kwargs...)` ([#38438])
 * New macros `@something` and `@coalesce` which are short-circuiting versions of `something` and `coalesce`, respectively ([#40729])
+* New function `redirect` for redirecting `stdin`, `stdout` and `stderr` ([#37978]).
```
Comment (barucden) [inline NEWS.md:49]: The function name has changed, right? The same applies for `doc/src/base/io-network.md`.

**Comment by jw3126** [comment]
Can this be merged? Would love to have this in 1.7.

**Comment by jw3126** [comment]
bump

**Comment by jw3126** [comment]
Bump can this please be merged before 1.7 feature freeze? This PR was ready from my side before 1.6 feature freeze and it would be frustrating to miss 1.7 as well.

**Review by fredrikekre — APPROVED** [review]

File: base/stream.jl [hunk 8]
```diff
@@ -1227,10 +1231,122 @@ Note that the direction of the stream is reversed.
...
+    if same_path(stderr, stdin)
+        throw(ArgumentError("stdin and stderr cannot be the same path"))
+    end
+    if same_path(stdout, stdin)
+        throw(ArgumentError("stdin and stdout cannot be the same path"))
+    end
+
+    new_in , close_in , old_in  = resolve(stdin , Base.stdin , "r")
+    new_out, close_out, old_out = resolve(stdout, Base.stdout, "w")
+    if same_path(stderr, stdout)
```
Comment (KristofferC) [inline base/stream.jl:1329]: Maybe use `Base.samefile`?

File: base/stream.jl [hunk 9]
```diff
@@ -1227,10 +1231,122 @@ Note that the direction of the stream is reversed.
...
+    end
+    function resolve(path::AbstractString, oldstream,mode)
+        (new=open(path, mode), close=true, old=oldstream)
+    end
+    function resolve(new, oldstream, mode)
+        (new=new, close=false, old=oldstream)
+    end
+
+    same_path(x, y) = false
+    same_path(x::AbstractString, y::AbstractString) = x == y
```
Comment (KristofferC) [inline base/stream.jl:1319]:
```suggestion
    same_path(x::AbstractString, y::AbstractString) = samefile(x, y)
```

maybe?

**Comment by jw3126** [comment]
Merge :)?

**Comment by jw3126** [comment]
@KristofferC can you merge this?

**Comment by jw3126** [comment]
I am glad, this could be merged before the feature freeze. Thanks, everybody!

**Pull request merged**